File: scripts/convert.py
import os
import sys
from bs4 import BeautifulSoup

def extract_headers_and_text(directory, filename):
    # Generate the full path to the HTML file
    html_file_path = os.path.join(directory, filename)
    print("Processing:", html_file_path)

    # Check if the file exists
    if not os.path.exists(html_file_path):
        print(f"File {html_file_path} does not exist.")
        return
    
    # Read the HTML content
    with open(html_file_path, 'r', encoding='utf-8') as f:
        html_content = f.read()
    
    # Parse the HTML content using BeautifulSoup
    soup = BeautifulSoup(html_content, 'html.parser')
    
    # Extract headers (h1, h2, h3) and their next valid text
    results = []
    for header in soup.find_all(['h1', 'h2', 'h3']):
        next_sibling = header.find_next_sibling()
        while next_sibling:
            if next_sibling.name:  # Ignore non-tag elements like comments or whitespace
                text = next_sibling.get_text(strip=True)
                if len(text) > 50:  # Adjust the length threshold as needed
                    results.append((header.name, header.get_text(strip=True), text))
                    break
            next_sibling = next_sibling.find_next_sibling()
    
    # Print or save the results
    for header_tag, header_text, text in results:
        print(f"{header_tag.upper()}: {header_text}")
        print(f"Text: {text}\n")
    
    return results

def main():
    if len(sys.argv) < 3:
        print("Usage: python convert.py <directory_path> <html_file1> <html_file2> ...")
        sys.exit(1)

    directory = sys.argv[1]
    html_files = sys.argv[2:]       

    for html_file in html_files:
        if html_file.endswith('.html'):
            extract_headers_and_text(directory, html_file)
        else:
            print(f"Skipping {html_file} (not an HTML file)")

if __name__ == "__main__":
    main()

File: scripts/convert.js
const fs = require('fs');
const path = require('path');
const { JSDOM } = require('jsdom');

function extractHeadersAndText(directory, filename) {
    const htmlFilePath = path.join(directory, filename);

    if (!fs.existsSync(htmlFilePath)) {
        return { error: `File ${htmlFilePath} does not exist.` };
    }

    const htmlContent = fs.readFileSync(htmlFilePath, 'utf-8');
    const dom = new JSDOM(htmlContent);
    const document = dom.window.document;
    const lines = htmlContent.split('\n');

    const getLineNumber = (element) => {
        const html = element.outerHTML;
        for (let i = 0; i < lines.length; i++) {
            if (lines[i].includes(html.substring(0, Math.min(100, html.length)))) {
                return i + 1;
            }
        }
        return -1;
    };

    const headerResults = [];
    const headerTexts = new Set();
    ['h1', 'h2', 'h3'].forEach(tag => {
        const headers = document.querySelectorAll(tag);
        headers.forEach(header => {
            const text = header.textContent.trim();
            if (text.length > 0) {
                headerResults.push({
                    headerTag: tag,
                    headerText: text,
                    headerLineNumber: getLineNumber(header),
                });
                headerTexts.add(text);
            }
        });
    });

    if (/contact/i.test(filename)) {
        return { headerResults: [], contentResults: [] };
      }

    const contentResults = [];
    document.querySelectorAll('*').forEach(element => {
        if (['SCRIPT', 'STYLE', 'NOSCRIPT'].includes(element.tagName)) return;

        const textContent = element.textContent.trim();

        
        if (
            textContent.length > 50 &&
            textContent.length <= 3000 && // filter out large blocks
            
            !headerTexts.has(textContent) // remove repeated headers
        ) {
            contentResults.push({
                text: textContent,
                lineNumber: getLineNumber(element),
            });
        }
    });

    return { headerResults, contentResults };
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 2) {
        console.log(JSON.stringify({ error: "Usage: node convert.js <directory_path> <html_file1> <html_file2> ..." }));
        process.exit(1);
    }

    const directory = args[0];
    const htmlFiles = args.slice(1);

    const results = htmlFiles.map(htmlFile => {
        if (htmlFile.endsWith('.html')) {
            return { file: htmlFile, ...extractHeadersAndText(directory, htmlFile) };
        }
        return { file: htmlFile, error: "Not an HTML file" };
    });

    console.log(JSON.stringify(results)); // Output results as JSON
}

if (require.main === module) {
    main();
}